packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts

```diff
@@ -0,0 +1,276 @@
+      logging: this.props.logging,
+      logGroup: this.props.logGroup,
+      role,
+      streamId: 'IntermediateS3',
+    }) ?? {};
+    const intermediateBucket = this.props.intermediateBucket ?? new s3.Bucket(scope, 'Intermediate Bucket');
+    intermediateBucket.grantReadWrite(role);
+    const intermediateS3Config: firehose.CfnDeliveryStream.S3DestinationConfigurationProperty = {
+      bucketArn: intermediateBucket.bucketArn,
+      roleArn: role.roleArn,
+      bufferingHints: createBufferingHints(bufferingInterval, bufferingSize),
+      encryptionConfiguration: createEncryptionConfig(role, this.props.encryptionKey),
+      cloudWatchLoggingOptions: intermediateLoggingOptions,
+      compressionFormat: compression?.value,
+    };
+    const bucketAccessRole = this.props.bucketAccessRole ?? new iam.Role(scope, 'Intermediate Bucket Access Role', {
+      assumedBy: new iam.ServicePrincipal('redshift.amazonaws.com'),
+    });
+    if (!this.props.bucketAccessRole && this.cluster instanceof redshift.Cluster) {
+      this.cluster.attachRole(bucketAccessRole);
+    }
+    intermediateBucket.grantRead(bucketAccessRole);
+
+    const { loggingOptions: redshiftLoggingOptions, dependables: redshiftLoggingDependables } = createLoggingOptions(scope, {
+      logging: this.props.logging,
+      logGroup: this.props.logGroup,
+      role,
+      streamId: 'Redshift',
+    }) ?? {};
+    const { backupConfig, dependables: backupDependables } = createBackupConfig(scope, role, this.props.s3Backup) ?? {};
+    return {
+      redshiftDestinationConfiguration: {
+        clusterJdbcurl: jdbcUrl,
+        copyCommand: {
+          dataTableName: tableName,
+          dataTableColumns: tableColumns.map(column => column.name).join(),
+          copyOptions: copyOptions + (compression === Compression.GZIP ? ' gzip' : ''),
+        },
+        password: user.password.toString(),
+        username: user.username.toString(),
+        s3Configuration: intermediateS3Config,
+        roleArn: role.roleArn,
+        cloudWatchLoggingOptions: redshiftLoggingOptions,
+        processingConfiguration: createProcessingConfig(scope, role, this.props.processor),
+        retryOptions: this.createRetryOptions(retryTimeout),
+        s3BackupConfiguration: backupConfig,
+        s3BackupMode: this.getS3BackupMode(),
+      },
+      dependables: [...(intermediateLoggingDependables ?? []), ...(redshiftLoggingDependables ?? []), ...(backupDependables ?? [])],
+    };
+  }
+
+  private createRetryOptions(retryTimeout?: cdk.Duration): firehose.CfnDeliveryStream.RedshiftRetryOptionsProperty | undefined {
+    return retryTimeout ? {
+      durationInSeconds: retryTimeout.toSeconds(),
+    } : undefined;
+  }
+
+  private getS3BackupMode(): string | undefined {
+    return (this.props.s3Backup?.bucket && !this.props.s3Backup.mode) || this.props.s3Backup?.mode === BackupMode.ALL
```

Comment:
This almost the same logic as: https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/s3-bucket.ts#L58 And I think they end up being equivalent. 

Should this be shared somewhere?

Reply:
That sounds good to me. A little bit of duplication is okay here in order to make it clear that each destination has full control over this part. The only way I can think of to make it clear and not have duplication would be to have a function called something like `getS3BackupModeForS3OrRedshift` which seems clunky and not flexible